File: LeetCode/May challenge/Day 15/Day_15.cpp
class Solution {
public:
    int maxSum(vector<int> &num)
    {
        int maxVal=num[0];
        int currMax=num[0];
        
        for(int i=1;i<num.size();i++)
        {
            currMax=max(num[i], currMax+num[i]);
            maxVal=max(maxVal,currMax);
        }
        return maxVal;
    }
    int maxSubarraySumCircular(vector<int>& A) {
        
        int sum1= maxSum(A);
        int totalsum=accumulate(A.begin(),A.end(),0);
        
        transform(A.begin(), A.end(), A.begin(), bind(multiplies<int>(),placeholders::_1, -1));
        
        int sum2=maxSum(A);
        
        if(sum2+totalsum > sum1 && sum2+totalsum !=0)
            return sum2+totalsum;
        else
            return sum1;
        
    }
};

File: LeetCode/May challenge/Day 15/Day_15.py
class Solution:
    def maxSubarraySumCircular(self, A: List[int]) -> int:
        n = len(A)
        if n == 0:
            return -1
        till_max = curr_max = till_min = curr_min = A[0]
        for num in A[1:]:
            curr_max = max(num, curr_max + num)
            till_max = max(till_max, curr_max)
            curr_min = min(num, curr_min + num)
            till_min = min(till_min, curr_min)
        if till_min == sum(A):
            return till_max
        return max(till_max, sum(A) - till_min)
